connectors/utils.py

```diff
@@ -113,8 +114,32 @@ def epoch_timestamp_zulu():
 
 def next_run(quartz_definition, now):
     """Returns the datetime in UTC timezone of the next run."""
-    cron_obj = QuartzCron(quartz_definition, now)
-    return with_utc_tz(cron_obj.next_trigger())
+    # Year is optional and is never present.
+    seconds, minutes, hours, day_of_month, month, day_of_week, year = (
+        quartz_definition.split(" ") + [None]
+    )[:7]
+
+    # Day of week is 1-7 starting from Sunday in Quartz and from Monday in regular Cron, adjust
+    # Days before: 1 - SUN, 2 - MON ... 7 - SAT
+    # Days after: 1 - MON, 2 - TUE ... 7 - SUN
+    if day_of_week.isnumeric():
+        day_of_week = (int(day_of_week) - 2) % 7 + 1
+
+    if not year:
+        year = "*"
+
+    # tzcron always expects year
+    repackaged_definition = (
+        f"{minutes} {hours} {day_of_month} {month} {day_of_week} {year}"
+    )
+
+    # ? comes from Quartz Cron, regular cron doesn't handle it well
+    repackaged_definition = repackaged_definition.replace("?", "*")
+
+    schedule = tzcron.Schedule(repackaged_definition, pytz.utc, now)
+
+    next_occurrence = next(schedule)
+    return with_utc_tz(next_occurrence)
```

Comment:
Just curious; is this converting a quartz expression into a cron expression? tzcron says it's also a quartz parser so I wonder why it's necessary to change anything. Is it just how it is?

Reply:
It's even funnier - it converts Quartz Cron into _whatever this cron in tzcron is_.

`tzcron` say that they support _most of_ Quartz definitions, but:

- Our Quartz cron has seconds as the first entry, for them first entry is always minutes
- As noted in comments, days of week are shifted (SUN-SAT for Quartz, MON-SUN for regular cron)
- "?" is not supported at all

🤷 

Hopefully once we switch to Connectors API for scheduling all of that can go away, cause Quartz is natively supported in Java, but for now there's no good way ahead that supports Quartz in Python, seems like